pczt/src/roles/signer/mod.rs

```diff
@@ -0,0 +1,302 @@
+};
+use zcash_protocol::consensus::BranchId;
+
+use crate::{common::Global, Pczt};
+
+use super::tx_extractor::determine_lock_time;
+
+const V5_TX_VERSION: u32 = 5;
+const V5_VERSION_GROUP_ID: u32 = 0x26A7270A;
+
+pub struct Signer {
+    global: Global,
+    transparent: transparent::pczt::Bundle,
+    sapling: sapling::pczt::Bundle,
+    orchard: orchard::pczt::Bundle,
+    /// Cached across multiple signatures.
+    tx_data: TransactionData<EffectsOnly>,
+    txid_parts: TxDigests<Blake2bHash>,
+    shielded_sighash: [u8; 32],
+    secp: secp256k1::Secp256k1<secp256k1::SignOnly>,
+}
+
+impl Signer {
+    /// Instantiates the Signer role with the given PCZT.
+    pub fn new(pczt: Pczt) -> Result<Self, Error> {
+        let Pczt {
+            global,
+            transparent,
+            sapling,
+            orchard,
+        } = pczt;
+
+        let transparent = transparent.into_parsed().map_err(Error::TransparentParse)?;
+        let sapling = sapling.into_parsed().map_err(Error::SaplingParse)?;
+        let orchard = orchard.into_parsed().map_err(Error::OrchardParse)?;
+
+        let tx_data = pczt_to_tx_data(&global, &transparent, &sapling, &orchard)?;
+        let txid_parts = tx_data.digest(TxIdDigester);
+
+        // TODO: Pick sighash based on tx version.
+        let shielded_sighash = v5_signature_hash(&tx_data, &SignableInput::Shielded, &txid_parts)
+            .as_ref()
+            .try_into()
+            .expect("correct length");
+
+        Ok(Self {
+            global,
+            transparent,
+            sapling,
+            orchard,
+            tx_data,
+            txid_parts,
+            shielded_sighash,
+            secp: secp256k1::Secp256k1::signing_only(),
+        })
+    }
+
+    /// Signs the transparent spend at the given index with the given spending key.
+    ///
+    /// It is the caller's responsibility to perform any semantic validity checks on the
+    /// PCZT (for example, comfirming that the change amounts are correct) before calling
+    /// this method.
+    pub fn sign_transparent(
+        &mut self,
+        index: usize,
+        sk: &secp256k1::SecretKey,
+    ) -> Result<(), Error> {
+        let input = self
+            .transparent
+            .inputs_mut()
+            .get_mut(index)
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        // TODO
+
+        input
+            .sign(index, &self.tx_data, &self.txid_parts, sk, &self.secp)
+            .map_err(Error::TransparentSign)?;
+
+        // Update transaction modifiability.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !0b0000_0001;
+        }
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
+            self.global.tx_modifiable &= !0b0000_0010;
+        }
+        if [SighashType::SINGLE, SighashType::SINGLE].contains(input.sighash_type()) {
```

Comment:
This was presumably supposed to be `[SighashType::SINGLE, SighashType::SINGLE_ANYONECANPAY]`. I think it's clearer like this though:
```suggestion
        if (input.sighash_type() & !SIGHASH_ANYONECANPAY) == SighashType::SINGLE {
```